### PR TITLE
Somewhat messy describeError implementation

### PR DESCRIPTION
**Pull request by Leo40Git**
The advanced mod resolver now knows how to handle missing dependencies, non-matching dependencies and breaking conflicts!

### CODE REVIEW

**Comment by AlexIIL** [comment]
Thanks! For some reason I didn't notice this earlier, sorry.

Unfortunately though the `describeError()` method has more complex semantics than with the simple mode (where each mod is mandatory, so any errors are always critical and must be reported). This list of causes is actually just reporting a single error.


For example if `buildcraft` jar-in-jars `libmultipart`, and `libmultipart` depends on `libnetworkstack` (but `libnetworkstack` isn't there) then:
* The root list will contain [ `MandatoryModIdDefinition(buildcraft)` ]
* The cause list will contain [ `OptionalModIdDefintion(libmultipart)`, `ModDep(buildcraft, libmultipart)`, `ModDep(libmultipart, libnetworkstack)` ]
(This is because it's not necessarily an issue if a mod is included but it's dependencies are missing, but only if it's actually required by a root/mandatory/user provided mod).

That's a fairly simple error though, as it can easily get much more complicated. For example:

* `buildcraft` depends on any version of `my-compat-layer`.
* `my-compat-layer` is made for specific versions of minecraft (a compatibility layer after all - this allows a single version of buildcraft to run on any version of minecraft)
* `buildcraft` includes both `my-compat-layer` for minecraft 1.16.x and 1.15.x (so that the same jar works on both)

However what if someone tries to install it into a snapshot? You'll get a list of causes that might want to be shown like this:

* Mandatory mod BuildCraft requires `my-compat-layer`, but no valid version could be found:
    * `my-compat-layer version 15` requires minecraft 1.15.x, but found minecraft `1.17-snapshot.20.45.a`
    * `my-compat-layer version 16` requires minecraft 1.16.x, but found minecraft `1.17-snapshot.20.45.a`
* You must install a version of `my-compat-layer` that works on minecraft `1.17-snapshot.20.45.a`, or downgrade it

But the list of roots will contain buildcraft and minecraft, and the list of causes will contain both `my-compat-layer` definitions, plus all 4 dependcies. (Hence why I haven't tried to write this yet, and why it probably needs a graph system to be made in order to make proper sense of the errors.

The other option is to special case the simpler options (probably by finding out if there's only 1 `ModDep` whose `validOptions` list is empty, and the rest are either `ModIdDefinition` or valid `ModDep`s, or by checking the size of the causes list is 1 or 2 and then listing every possibility that makes sense)

Basically I hope you can see why I haven't attempted to implement this yet :p

**Comment by Leo40Git** [comment]
...oh. That's a lot more complex than I was expecting.
I'll see what I can do about implementing this better, though...

**Comment by AlexIIL** [comment]
Can I merge this in as-is? While I don't think this works correctly I think your error messages are better than what I would have made?

**Comment by Leo40Git** [comment]
Oh, sure!
Sorry for the radio silence, I wasn't able to figure out how to do it properly...

**Comment by AlexIIL** [comment]
Fair enough :p
